pkg/goDB/conditions/node/filter_direction_test.go

```diff
@@ -0,0 +1,132 @@
+		expectedCondition: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"},
+		expectedFilter:    types.Counters.IsBidirectional, expectedErr: nil},
+	// invalid filter (2 traffic filters provided)
+	{node: andNode{left: conditionNode{attribute: "dir", comparator: "=", value: "uni"},
+		right: conditionNode{attribute: "dir", comparator: "=", value: "bi"}},
+		expectedCondition: nil,
+		expectedFilter:    nil, expectedErr: errMultipleDirectionFilterConditions},
+	// invalid filter (2 traffic filters provided, of which one is nested)
+	{node: andNode{left: conditionNode{attribute: "dir", comparator: "=", value: "uni"},
+		right: andNode{left: conditionNode{attribute: "dip", comparator: "=", value: "127.0.0.1"}, right: conditionNode{attribute: "dir", comparator: "=", value: "bi"}}},
+		expectedCondition: nil,
+		expectedFilter:    nil, expectedErr: errMultipleDirectionFilterConditions},
+	// simple condition without any traffic filter => valid
+	{node: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"},
+		expectedCondition: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"},
+		expectedFilter:    nil, expectedErr: nil},
+	// disjunctive condition with traffic filter => invalid
+	{node: orNode{left: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"}, right: conditionNode{attribute: "dir", comparator: "=", value: "in"}},
+		expectedCondition: nil,
+		expectedFilter:    nil, expectedErr: errMisplacedDirectionFilterCondition},
+	// conjunctive condition without any traffic filter => valid
+	{node: andNode{left: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"}, right: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"}},
+		expectedCondition: andNode{left: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"}, right: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"}},
+		expectedFilter:    nil, expectedErr: nil},
+	// conjunctive condition with misplaced traffic filter => invalid
+	{node: andNode{left: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"},
+		right: orNode{left: conditionNode{attribute: "sip", comparator: "=", value: "127.0.0.1"}, right: conditionNode{attribute: "dir", comparator: "=", value: "in"}}},
+		expectedCondition: nil,
+		expectedFilter:    nil, expectedErr: errMisplacedDirectionFilterCondition},
+}
+
+func TestSplitOffDirectionFilter(t *testing.T) {
+	for _, test := range splitOffDirectionFilterTest {
+		condition, valFilterNode, err := splitOffDirectionFilter(test.node)
+		if err != nil && err.Error() != test.expectedErr.Error() {
+			t.Fatalf("Expected error: %s Actual error: %s\n", test.expectedErr, err)
+		}
+
+		if err == nil && reflect.ValueOf(valFilterNode.ValFilter) != reflect.ValueOf(test.expectedFilter) {
```

Comment:
AFAIK the use of `reflect.ValueOf(fn)` for function equality produces (or rather: can produce) undefined behavior, see e.g. https://stackoverflow.com/questions/34901307/how-to-compare-2-functions-in-go (I couldn't find the official spec / statement at first glance, but I'm pretty sure the general statement about comparing functions is: don't)...
Maybe `testify` comes with something, not sure...

Reply:
I've also seen this post and agree with you. 
It seems that also `testify` cannot help in this case, as it uses reflection underneath (stretchr/testify/issues/159). 
I've double-checked the tests and they work as expected (different functions yield different addresses). I would suggest to leave it as is for now and investigate further if problems arise due to these test cases.
One possible thing we could do is to include more details about the filter inside the `ValFilterNode` to circumvent reflection. IMO we shouldn't do this because we would only check some filter description instead of the actual filter (and we would maintain state that is only used for testing).